src/gf.c

```diff
@@ -1496,118 +1505,287 @@ static void parameters_to_closureenv(jl_value_t *ast, jl_svec_t *tvars)
+            jl_trampoline_compile_linfo(li, 1);
+            assert(li->functionID > 0);
+            continue;
+        }
+
+        jl_function_t *func = (jl_function_t*)thunk;
+        jl_methlist_t *meth = (jl_methlist_t*)jl_cellref(found, found_i + 1);
+        assert(!func->linfo);
+
+        if (jl_is_leaf_type((jl_value_t*)meth->sig)) {
+            // usually can create a specialized version of the function,
+            // if the signature is already a leaftype
+            jl_function_t *spec = jl_get_specialization(func, meth->sig);
+            if (spec && !jl_has_typevars((jl_value_t*)meth->sig)) {
+                // replace unspecialized func with specialized version
+                // if there are no bound type vars (e.g. `call{K,V}(Dict{K,V})` vs `call(Dict)`)
+                // that might cause a different method to match at runtime
+                meth->func = spec;
+                jl_gc_wb(meth, spec);
                 continue;
             }
         }
-        func = m->func->linfo->unspecialized;
-        if (func == NULL) {
-            func = jl_instantiate_method(m->func, jl_emptysvec);
-            if (func->env != (jl_value_t*)jl_emptysvec)
-                func->env = NULL;
-            m->func->linfo->unspecialized = func;
-            jl_gc_wb(m->func->linfo, func);
+
+        if (jl_is_typevar(meth->tvars)) {
+            // f{T<:Union{...}}(...) is a common pattern
+            // and expanding the Union may give a leaf function
+            jl_tvar_t *tv = (jl_tvar_t*)meth->tvars;
+            if (jl_is_uniontype(tv->ub)) {
+                int complete = 1;
+                jl_uniontype_t *ub = (jl_uniontype_t*)tv->ub;
+                size_t i, l = jl_svec_len(ub->types);
+                for (i = 0; i < l + 1; i++) { // add Union{} to the end of the list, since T<:Union{} is always a valid option
+                    jl_value_t *ty = (i == l ? jl_bottom_type : jl_svecref(ub->types, i));
+                    if (i == l || jl_is_leaf_type(ty)) {
+                        jl_value_t *env[2] = {(jl_value_t*)tv, ty};
+                        jl_value_t *sig;
+                        JL_TRY {
+                            sig = (jl_value_t*)
+                                jl_instantiate_type_with((jl_value_t*)meth->sig, env, 1);
+                        }
+                        JL_CATCH {
+                            continue; // sigh, we found an invalid type signature. should we warn the user?
+                        }
+                        assert(jl_is_tuple_type(sig));
+                        if (sig == jl_bottom_type || tupletype_any_bottom(sig)) {
+                            continue; // signature wouldn't be callable / is invalid -- skip it
+                        }
+                        if (jl_is_leaf_type(sig)) {
+                            if (jl_get_specialization(func, (jl_tupletype_t*)sig)) {
+                                continue;
```

Comment:
TODO: may need to consider `jl_has_typevars((jl_value_t*)meth->sig)` like above.